sycl/test/reduction/reduction_usm.cpp

```diff
@@ -0,0 +1,77 @@
+// UNSUPPORTED: cuda
+// Reductions use work-group builtins not yet supported by CUDA.
+
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// This test performs basic checks of parallel_for(nd_range, reduction, func)
+// with reductions initialized with USM var.
+
+#include "reduction_utils.hpp"
+#include <CL/sycl.hpp>
+#include <cassert>
+
+using namespace cl::sycl;
+
+template <typename T, int Dim, class BinaryOperation>
+class SomeClass;
+
+template <typename T, int Dim, class BinaryOperation>
+void test(T Identity, size_t WGSize, size_t NWItems) {
+  // Initialize.
+  T CorrectOut;
+  BinaryOperation BOp;
+
+  buffer<T, 1> InBuf(NWItems);
+  initInputData(InBuf, CorrectOut, Identity, BOp, NWItems);
+
+  T ReduVar = Identity;
```

Comment:
`ReduVar` isn't a USM allocation.

I like what you've implemented, but it isn't performing reductions using a variable allocated in USM -- it's performing reductions using a variable allocated on the host.  This may be an important use-case, and I wonder if we should update the reduction proposal to make it legal... any thoughts, @rolandschulz @jasonsewall-intel?

The use-case we were targeting with USM would look something like the following:
```c++
queue Q;
T* ReduVar = malloc_device(sizeof(T), q);
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVar, Identity, BOp);
  ...
});
```

Reply:
Good catch, I misunderstood the reduction proposal somehow.
I can quickly update the patch to make reduction_impl constructor to accept 'T *ReduVarPtr' instead of 'T &ReduVar'.

Later, if there is the agreement regarding passing reference to Var, I can add 2 more constructors (from the current patch) to reduction (to support variant 2 below).

```
//Variant1
queue Q;
T* ReduVarPtr = malloc_device(sizeof(T), q);
*ReduVarPtr = Identity;
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVarPtr, Identity, BOp);
  ...
});
//Variant2
queue Q;
T ReduVar = Identity;
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVar, Identity, BOp);
  ...
});
```

BTW, fix me if I am wrong here: handler::copy(accessor, ptr) does not require 'ptr' to be USM pointer, right? If So, then in the current/proposed patch USM'ness is just not required at all.

Reply:
John, I did the additional fix. Please see the 2nd commit in this PR.
Thank you!

Reply:
I'm not sure about this.  If `handler::copy(accessor, ptr)` runs on the host, it won't be able to copy from a device pointer into an accessor.

@jbrodman: What is the right way to copy from a USM pointer into an accessor?

Reply:
I'm still slightly surprised that `CopyHandler.copy(Acc, Redu.getUSMPointer());` does the right thing here, but since the `malloc_device` test is passing I'll mark this resolved.